File: framework/headless/action.go
package headless

import (
	"strconv"
	"strings"

	"github.com/DonaldMurillo/gofastr/core-ui/html"
	"github.com/DonaldMurillo/gofastr/core/render"
)

// The optimistic mutation buttons, carrying GoFastr's
// optimistic-action contract so the framework's own runtime drives
// them: click paints the committed label immediately, the endpoint is
// fired, and a non-2xx rolls the label back. The HTTP response is the
// only authority — never a pushed event.
//
// Two components, one lifecycle. OptimisticAction commits once and
// stays committed ("Follow"); ToggleAction flips a binary the server
// owns ("Watch"/"Watching"), can ship already committed, and knows
// about mutex groups and untoggle.
//
// What the framework's runtime does NOT do is tell anyone: on failure
// it plays a shake and announces nothing. Both components here render
// a status span — role="status", polite, visually hidden — and put
// the failure sentence on the root as data-hui-action-failed, which is
// everything a runtime module that binds the hooks needs to turn the
// framework's bubbled rolled-back event into words.

// Action parts. The idle and done labels are parts of their own
// because a skin may want to weigh one differently from the other —
// a committed toggle that reads heavier than its idle sibling. The
// status span is PartVisuallyHidden because it must be read and must
// not be seen, the same rule as every other off-screen sentence in
// this package.
const (
	PartActionIdle Part = "action-idle"
	PartActionDone Part = "action-done"
)

// The failed-mutation sentence lives in Strings.ActionFailed; a
// sentence and not a bare "Error" is the point — see that field.

// checkActionEndpoint refuses the two endpoints that look wired and
// are not: none at all, and one the runtime will decline to fetch.
// The framework's origin check drops a cross-origin URL silently —
// the click does nothing — so the same-origin rule is enforced here,
// where the mistake is a panic at render time rather than a dead
// button in production.
func checkActionEndpoint(component, what, endpoint string) {
	if endpoint == "" {
		panic("headless: " + component + " requires " + what)
	}
	checkSameOrigin(component, what, endpoint)
}

// actionMethod canonicalises the method and returns the value to
// render: nothing for POST, which is the default the runtime already
// assumes, so an unstated method and an omitted attribute are the
// same request. Anything but the four verbs the runtime will send is
// refused rather than rendered into a fetch that cannot happen.
func actionMethod(component, method string) string {
	switch method {
	case "", "POST":
		return ""
	case "PUT", "PATCH", "DELETE":
		return method
	default:
		panic("headless: " + component + " Method must be POST, PUT, PATCH or DELETE, not " + method)
	}
}

// checkLabelsDiffer refuses two labels that read the same. The flip
// between them is the state change; with identical labels the only
// remaining signal is colour, which is the one signal this system
// never lets stand alone (WCAG 1.4.1). Whitespace and case are not a
// difference anyone hears.
func checkLabelsDiffer(component, idle, done string) {
	if strings.EqualFold(strings.TrimSpace(idle), strings.TrimSpace(done)) {
		panic("headless: " + component + " labels must differ — the flip from one to the other is the state change, and " +
			"two labels reading " + strconv.Quote(idle) + " leave colour as the only signal")
	}
}

// safeActionExtras is Safe with the prefixes the lifecycle owns added:
// a caller may not forge a data-hui-* hook any more than a data-fui-*
// one, and the names Safe takes exactly — type, disabled, data-state,
// aria-busy, aria-pressed — are the ones the runtime rewrites as the
// mutation moves. An extra that won any of them would desynchronise
// the button from its own lifecycle.
func safeActionExtras(extra html.Attrs) html.Attrs {
	return Safe(extra, "type", "disabled", "data-state", "aria-busy", "aria-pressed")
}

// OptimisticActionProps is a button that commits once: the success
// label is painted on click, the endpoint fired, and a non-2xx rolls
// everything back with a shake.
type OptimisticActionProps struct {
	// Endpoint is the URL the click fires against. Required, and
	// same-origin (it must start with "/").
	Endpoint string
	// Method defaults to POST and may be PUT, PATCH or DELETE.
	Method string
	// IdleLabel is the text at rest. Required.
	IdleLabel string
	// SuccessLabel is the text painted the moment the button is
	// clicked. Required — and it must say something the idle label
	// does not, because the label flip IS the state change; colour
	// alone is never allowed to carry it.
	SuccessLabel string
	// IdleIcon and DoneIcon are decoration beside each label. Each
	// span carries its own so the flip swaps glyph and word together.
	IdleIcon render.HTML
	DoneIcon render.HTML
	// Variant and Size are skin vocabulary, passed through so the
	// skin can look up "<part>--<variant>". The structure does not
	// care.
	Variant string
	Size    string
	// Disabled is the state at render time. The runtime manages it
	// from the first click: pending disables, settlement re-enables.
	Disabled bool
	// FailedText is what the status span announces when the server
	// refuses. Defaults to "Could not save. Try again."
	FailedText string

	ID         string
	ExtraAttrs html.Attrs

	// Parts: attrs only. Nothing is fillable — the two labels are
	// the state, and a slot that replaced one could make the button
	// say it did something it did not.
	Parts Parts
	// Strings are the strings this component says. Nil means the English
	// defaults; a layer above sets them from the request's language.
	Strings *Strings
}

// ToggleActionProps is the three-state cousin: idle → pending →
// committed, and back, with the initial state shipped from the server.
type ToggleActionProps struct {
	// Endpoint is the URL hit when toggling idle → committed.
	// Required, and same-origin.
	Endpoint string
	// Method defaults to POST and may be PUT, PATCH or DELETE. It
	// applies to both the commit and the untoggle request.
	Method string
	// IdleLabel is the text in the un-committed state. Required.
	IdleLabel string
	// CommittedLabel is the text while committed. Required, and it
	// must differ from IdleLabel: the flip is the signal.
	CommittedLabel string
	// IdleIcon and DoneIcon are decoration beside each label.
	IdleIcon render.HTML
	DoneIcon render.HTML
	// Committed is the state the server knows: render it true when
	// the action is already active, so first paint matches server
	// state instead of flashing through the wrong label.
	Committed bool
	// Group, when set, joins this button to a client-side mutex:
	// committing any button with the same key reverts the
	// previously-committed sibling, with no second request.
	Group string
	// AllowUntoggle lets a click on a committed button revert it to
	// idle. Without it (and without UntoggleEndpoint) the button is
	// sticky once committed, like OptimisticAction.
	AllowUntoggle bool
	// UntoggleEndpoint is the URL hit when reverting committed →
	// idle. Setting it implies AllowUntoggle; when empty the revert
	// flips locally with no request.
	UntoggleEndpoint string
	// Variant and Size are skin vocabulary.
	Variant string
	Size    string
	// Disabled is the state at render time.
	Disabled bool
	// FailedText is what the status span announces on a failure.
	FailedText string

	ID         string
	ExtraAttrs html.Attrs

	// Parts: attrs only, for the same reason as
	// OptimisticActionProps.
	Parts Parts
	// Strings are the strings this component says. Nil means the English
	// defaults; a layer above sets them from the request's language.
	Strings *Strings
}

// action is the half of the contract the two buttons share, so the
// shared half is rendered once. The differences are attributes and
// initial state, never structure: the runtime expects one shape from
// both, and a component that drifted would work until the page it was
// on was read by the other one's test.
type action struct {
	comp string // data-fui-comp: which runtime module drives it
	// wire is the attribute prefix that module reads its endpoint and
	// method from: "data-fui-optimistic" or "data-fui-toggle".
	wire     string
	idleAttr string // presence hook the runtime's setState flips hidden between
	doneAttr string

	endpoint string
	method   string // "" means POST and renders no attribute

	idleLabel string
	doneLabel string
	idleIcon  render.HTML
	doneIcon  render.HTML

	// state is "idle" or "committed". Pending is never shipped: it is
	// painted after a click and cleared when the response lands, and a
	// server that rendered it would send a button disabled and busy
	// before anyone had touched it.
	state string
	// pressed is the aria-pressed value, "" when the button is not a
	// toggle: OptimisticAction commits once and is not a pressed
	// control, so it claims no pressed state.
	pressed string

	group         string
	allowUntoggle bool
	untoggle      string

	variant    string
	size       string
	disabled   bool
	failedText string
	id         string
	extra      html.Attrs
	parts      Parts
	strings    *Strings
}

// OptimisticAction renders the button. The framework's runtime binds
// on [data-fui-comp="ui-optimistic-action"]; everything it needs —
// endpoint, method, both labels — is in the markup below.
func OptimisticAction(p OptimisticActionProps, s Skin) render.HTML {
	checkActionEndpoint("OptimisticAction", "Endpoint", p.Endpoint)
	if p.IdleLabel == "" {
		panic("headless: OptimisticAction requires IdleLabel")
	}
	if p.SuccessLabel == "" {
		panic("headless: OptimisticAction requires SuccessLabel")
	}
	checkLabelsDiffer("OptimisticAction", p.IdleLabel, p.SuccessLabel)
	return renderAction(action{
		comp:       "ui-optimistic-action",
		wire:       "data-fui-optimistic",
		idleAttr:   "data-fui-optimistic-idle",
		doneAttr:   "data-fui-optimistic-success",
		endpoint:   p.Endpoint,
		method:     actionMethod("OptimisticAction", p.Method),
		idleLabel:  p.IdleLabel,
		doneLabel:  p.SuccessLabel,
		idleIcon:   p.IdleIcon,
		doneIcon:   p.DoneIcon,
		state:      "idle",
		variant:    p.Variant,
		size:       p.Size,
		disabled:   p.Disabled,
		failedText: p.FailedText,
		id:         p.ID,
		extra:      p.ExtraAttrs,
		parts:      p.Parts,
		strings:    p.Strings,
	}, s)
}

// ToggleAction renders the button. The framework's runtime binds on
// [data-fui-comp="ui-toggle-action"], ships nothing itself but the
// initial state below, and mirrors committed onto aria-pressed from
// then on.
func ToggleAction(p ToggleActionProps, s Skin) render.HTML {
	checkActionEndpoint("ToggleAction", "Endpoint", p.Endpoint)
	if p.IdleLabel == "" {
		panic("headless: ToggleAction requires IdleLabel")
	}
	if p.CommittedLabel == "" {
		panic("headless: ToggleAction requires CommittedLabel")
	}
	checkLabelsDiffer("ToggleAction", p.IdleLabel, p.CommittedLabel)
	if p.UntoggleEndpoint != "" {
		checkActionEndpoint("ToggleAction", "UntoggleEndpoint", p.UntoggleEndpoint)
	}
	state, pressed := "idle", "false"
	if p.Committed {
		state, pressed = "committed", "true"
	}
	return renderAction(action{
		comp:      "ui-toggle-action",
		wire:      "data-fui-toggle",
		idleAttr:  "data-fui-toggle-idle",
		doneAttr:  "data-fui-toggle-committed",
		endpoint:  p.Endpoint,
		method:    actionMethod("ToggleAction", p.Method),
		idleLabel: p.IdleLabel,
		doneLabel: p.CommittedLabel,
		idleIcon:  p.IdleIcon,
		doneIcon:  p.DoneIcon,
		state:     state,
		pressed:   pressed,
		group:     p.Group,
		// Setting an untoggle endpoint is itself the request to be
		// untoggleable, the same implication the framework makes.
		allowUntoggle: p.AllowUntoggle || p.UntoggleEndpoint != "",
		untoggle:      p.UntoggleEndpoint,
		variant:       p.Variant,
		size:          p.Size,
		disabled:      p.Disabled,
		failedText:    p.FailedText,
		id:            p.ID,
		extra:         p.ExtraAttrs,
		parts:         p.Parts,
		strings:       p.Strings,
	}, s)
}

// renderAction draws the button both components are: the lifecycle
// attributes the framework's runtime reads, the announcement hooks
// ours writes to, and two label spans of which exactly the one
// matching the shipped state is visible.
func renderAction(a action, s Skin) render.HTML {
	own := Merge(safeActionExtras(a.extra), Attrs(map[string]string{
		"id":            a.id,
		"type":          "button",
		"data-fui-comp": a.comp,
	}))
	own[a.wire+"-endpoint"] = a.endpoint
	if a.method != "" {
		own[a.wire+"-method"] = a.method
	}
	if a.group != "" {
		own["data-fui-toggle-group"] = a.group
	}
	if a.allowUntoggle {
		own["data-fui-toggle-allow-untoggle"] = "true"
	}
	if a.untoggle != "" {
		own["data-fui-toggle-untoggle-endpoint"] = a.untoggle
	}
	own["data-state"] = a.state
	if a.pressed != "" {
		own["aria-pressed"] = a.pressed
	}
	// The sentence, on the root: the runtime's rolled-back listener
	// finds the button, and the span it writes to is inside it, so
	// the announcement never has to reach for anything the flip
	// already moved.
	own["data-hui-action-failed"] = orDefault(a.failedText, a.strings.Resolve().ActionFailed)
	Mark(own, "data-hui-action")
	Flag(own, "disabled", a.disabled)
	if cls := s.Variant(PartRoot, a.variant); cls != "" {
		own["class"] = cls
	}
	if cls := s.Variant(PartRoot, a.size); cls != "" {
		own["class"] = joinClasses(own["class"], cls)
	}

	b := a.parts.Box(s)
	kids := []render.HTML{
		actionSpan(b, PartActionIdle, a.idleAttr, a.state == "committed", a.idleLabel, a.idleIcon),
		actionSpan(b, PartActionDone, a.doneAttr, a.state != "committed", a.doneLabel, a.doneIcon),
		// The accessibility the framework's runtime does not ship.
		// A polite status region — role=status already means polite,
		// and stating it twice can announce twice: the flip is
		// visible, and this is how the rollback reaches anyone who
		// cannot see it.
		b.El("span", PartVisuallyHidden, Mark(Attrs(map[string]string{
			"role": "status",
		}), "data-hui-action-status")),
	}
	return b.El("button", PartRoot, own, kids...)
}

// actionSpan renders one of the two labels. hidden decides which of
// the pair carries it: the runtime flips that attribute and nothing
// else, so the server ships exactly the opposite pair and the first
// paint already agrees with the state attribute above it.
func actionSpan(b Box, part Part, hook string, hidden bool, label string, icon render.HTML) render.HTML {
	attrs := html.Attrs{hook: ""}
	if hidden {
		Mark(attrs, "hidden")
	}
	kids := make([]render.HTML, 0, 2)
	if icon != "" {
		kids = append(kids, b.El("span", PartIcon,
			Attrs(map[string]string{"aria-hidden": "true"}), icon))
	}
	kids = append(kids, render.Text(label))
	return b.El("span", part, attrs, kids...)
}

func init() {
	Register(Spec{
		Name:    "OptimisticAction",
		Anatomy: []Part{PartRoot, PartIcon, PartActionIdle, PartActionDone, PartVisuallyHidden},
		Hooks:   []string{"data-hui-action", "data-hui-action-failed", "data-hui-action-status"},
		WithParts: func(s Skin, parts Parts) render.HTML {
			return OptimisticAction(OptimisticActionProps{
				Endpoint: "/follow", IdleLabel: "Follow", SuccessLabel: "Following",
				Parts: parts,
			}, s)
		},
		Cases: func(k Kit) []Case {
			s := k.Skin
			return []Case{{
				Name: "default",
				Why: "both labels ship in the SSR so the optimistic flip needs no round-trip, the method is " +
					"absent because POST is the runtime's default, and the status span exists because the " +
					"framework's shake announces nothing",
				HTML: OptimisticAction(OptimisticActionProps{
					Endpoint: "/apps/blog/restart", IdleLabel: "Restart now", SuccessLabel: "Restarting",
					Variant: "primary",
				}, s),
			}, {
				Name: "with icons",
				Why: "each label carries its own icon so the flip swaps glyph and word together, and both are " +
					"aria-hidden because the words are the meaning — a check mark nobody can hear says nothing",
				HTML: OptimisticAction(OptimisticActionProps{
					Endpoint: "/apps/blog/pin", IdleLabel: "Pin", SuccessLabel: "Pinned",
					IdleIcon: SpecimenGlyph, DoneIcon: SpecimenGlyph, Variant: "secondary",
				}, s),
			}, {
				Name: "delete, not post",
				Why: "a method other than POST is published as an attribute, because the runtime defaults to " +
					"POST and an unstated DELETE would fire the wrong request entirely",
				HTML: OptimisticAction(OptimisticActionProps{
					Endpoint: "/webhooks/9", Method: "DELETE",
					IdleLabel: "Remove webhook", SuccessLabel: "Removed", Variant: "danger",
				}, s),
			}, {
				Name: "disabled",
				Why: "a disabled optimistic button still ships both labels and its announcement hook — " +
					"disabling is a state of the click, not of the contract, and the runtime re-enables " +
					"the button itself while a mutation is in flight",
				HTML: OptimisticAction(OptimisticActionProps{
					Endpoint: "/apps/blog/redeploy", IdleLabel: "Redeploy", SuccessLabel: "Redeploying",
					Variant: "ghost", Disabled: true,
				}, s),
			}}
		},
	})

	Register(Spec{
		Name:    "ToggleAction",
		Anatomy: []Part{PartRoot, PartIcon, PartActionIdle, PartActionDone, PartVisuallyHidden},
		Hooks:   []string{"data-hui-action", "data-hui-action-failed", "data-hui-action-status"},
		WithParts: func(s Skin, parts Parts) render.HTML {
			return ToggleAction(ToggleActionProps{
				Endpoint: "/watch", IdleLabel: "Watch", CommittedLabel: "Watching",
				Parts: parts,
			}, s)
		},
		Cases: func(k Kit) []Case {
			s := k.Skin
			return []Case{{
				Name: "idle",
				Why: "aria-pressed false and the committed span hidden: first paint matches the server's " +
					"answer, and the runtime only ever flips what is already there",
				HTML: ToggleAction(ToggleActionProps{
					Endpoint: "/apps/blog/watch", IdleLabel: "Watch", CommittedLabel: "Watching",
					Variant: "primary",
				}, s),
			}, {
				Name: "committed at load",
				Why: "the server knows: aria-pressed is true and it is the idle span that is hidden, so a page " +
					"that loads already-watching does not flash through the wrong label first — and each label " +
					"carries its own aria-hidden icon, so the flip swaps glyph and word together",
				HTML: ToggleAction(ToggleActionProps{
					Endpoint: "/apps/blog/watch", IdleLabel: "Watch", CommittedLabel: "Watching",
					IdleIcon: SpecimenGlyph, DoneIcon: SpecimenGlyph,
					Committed: true, Variant: "secondary",
				}, s),
			}, {
				Name: "two in a mutex group",
				Why: "committing one member of a group revokes the other with no second request, so a plan " +
					"picker reads as one choice rather than three independent buttons that can all be on",
				HTML: group(
					ToggleAction(ToggleActionProps{
						Endpoint: "/plan/starter", IdleLabel: "Starter", CommittedLabel: "Starter ✓",
						Group: "plan", Committed: true, Variant: "ghost",
					}, s),
					ToggleAction(ToggleActionProps{
						Endpoint: "/plan/pro", IdleLabel: "Pro", CommittedLabel: "Pro ✓",
						Group: "plan", Variant: "ghost",
					}, s)),
			}, {
				Name: "untoggle, with its own endpoint",
				Why: "a committed button that can be clicked again publishes the untoggle endpoint and ships " +
					"committed — the only state from which untoggle is reachable — so the revert is a request " +
					"the server can refuse, not just a local flip",
				HTML: ToggleAction(ToggleActionProps{
					Endpoint: "/plan/pro", UntoggleEndpoint: "/plan/none",
					IdleLabel: "Pro", CommittedLabel: "Pro ✓", Committed: true,
					AllowUntoggle: true, Variant: "primary",
				}, s),
			}}
		},
	})
}

File: core-ui/runtime/src/action.js
// action.js: the optimistic mutation machine, written once.
//
// OptimisticAction and ToggleAction were one state machine implemented
// twice (src/optimisticaction.js, src/toggleaction.js, both now in
// framework/ui as adapters over this module). What lives here is the
// half they shared: a same-origin mutation request with the CSRF
// header and the invalidation hook, the idle → pending → committed →
// error state flip, hidden swapped between the two label parts,
// aria-busy and disabled while pending, and an event at each step.
// Attribute names are the owner's business: an adapter reads its own
// markup and calls bind with a spec, so this module reads no data-fui-*
// attribute at all and scans nothing.
//
// It has no marker, so nothing loads it on its own: a module that
// needs it declares registry.Requires("action") and the loader has it
// registered before the dependent evaluates. window.__gofastr.action
// is also a public API (request, bind) for surfaces with no adapter.
(() => {
  'use strict';
  const NS = window.__gofastr = window.__gofastr || {};

  // request performs the mutation. It resolves true on a 2xx, false on
  // a non-2xx, a network error, a refused origin or an empty URL, and
  // never throws, so the caller's settlement is always exactly one of
  // the two outcomes. A 2xx is handed to the kernel's invalidation
  // hook first (X-Gofastr-Invalidate), the way src/rpc.js does, so an
  // action refreshes the screens it changed.
  function request(url, method) {
    if (!url) return Promise.resolve(false);
    if (!NS._originOK || !NS._originOK(url)) return Promise.resolve(false);
    // CSRF: forward the page's <meta name="csrf-token"> as X-CSRF-Token
    // on every state-changing fetch, the contract hard rule 10 states.
    // Apps verify the token server-side; the runtime makes the value
    // reachable so each mutation surface does not re-implement the
    // lookup.
    const headers = { Accept: 'application/json' };
    const meta = document.querySelector('meta[name="csrf-token"]');
    const token = meta && meta.getAttribute('content');
    if (token) headers['X-CSRF-Token'] = token;
    return fetch(url, {
      method: (method || 'POST').toUpperCase(),
      credentials: 'same-origin',
      headers,
    }).then((res) => {
      if (!res.ok) return false;
      if (NS._inval) NS._inval(res);
      return true;
    }, () => false);
  }

  // Per-element state. The spec is kept in a WeakMap so a button that
  // leaves the DOM takes its state with it, and binding twice is a
  // no-op through the bound WeakSet: an adapter's scan and the kernel's
  // insertion pass both reach the same element.
  const specs = new WeakMap();
  const bound = new WeakSet();
  // Group members, keyed by group name. A Map of Sets rather than a
  // document scan on every commit: the registry is the truth about what
  // is bound. Members whose elements left the document are pruned at
  // iteration time (an island swap replaces a whole group at once; the
  // stale members must not be revoked-or-kept by accident) and on
  // gofastr:navigate (see the handler below), so a group navigated away
  // from is not held forever.
  const groups = new Map();

  // setState paints one state. idle and done are the two label parts
  // (elements, from the spec): done is shown for pending and committed
  // — the optimistic flip paints the success label the moment the
  // click lands — and idle otherwise. aria-busy lives on pending only:
  // the committed state is final but focusable, so an app can replace
  // the DOM with an undo affordance. The element's disabled attribute
  // is never touched: pending already ignores clicks (see bind), a
  // disabled button drops keyboard focus to the body the moment it is
  // disabled, and disabled has other owners (a hidden conditional
  // region disables its controls) whose decision a settlement must not
  // undo. When the spec asked for pressed, aria-pressed mirrors
  // committed, which is the pressed-toggle convention.
  function setState(el, spec, state) {
    el.setAttribute('data-state', state);
    if (spec.idle && spec.done) {
      if (state === 'pending' || state === 'committed') {
        spec.idle.setAttribute('hidden', '');
        spec.done.removeAttribute('hidden');
      } else {
        spec.idle.removeAttribute('hidden');
        spec.done.setAttribute('hidden', '');
      }
    }
    if (state === 'pending') {
      el.setAttribute('aria-busy', 'true');
    } else {
      el.removeAttribute('aria-busy');
    }
    if (spec.pressed) el.setAttribute('aria-pressed', state === 'committed' ? 'true' : 'false');
  }

  // revokeGroupSiblings flips every other committed member of the group
  // back to idle, with no second request: the server stays the source
  // of truth and a later navigation refreshes from it. It returns the
  // members it revoked so a failed commit can put them back: the
  // server never accepted the new member, so the old one is still the
  // committed one. Members whose elements left the document are pruned
  // here, on iteration, which is the only pass that would otherwise
  // hold them.
  function revokeGroupSiblings(group, except) {
    const revoked = [];
    const members = groups.get(group);
    if (!members) return revoked;
    for (const other of members) {
      if (!other.isConnected) {
        members.delete(other);
        continue;
      }
      if (other === except) continue;
      if (other.getAttribute('data-state') === 'committed') {
        setState(other, specs.get(other), 'idle');
        revoked.push(other);
      }
    }
    if (members.size === 0) groups.delete(group);
    return revoked;
  }

  // stillGrouped reports whether el is still a member of its group in
  // the document: a settlement that arrives after an island swap
  // replaced the group must not revoke the new page's committed member
  // on behalf of a button that is no longer there.
  function stillGrouped(spec, el) {
    if (!spec.group || !el.isConnected) return false;
    const members = groups.get(spec.group);
    return !!members && members.has(el);
  }

  // bind attaches the lifecycle, once per element.
  //
  // spec: endpoint (the commit URL), method (default POST), idle and
  // done (the two label parts, elements), and optionally group (a
  // mutex key), untoggle and pressed. untoggle arms the revert path:
  // a string is the untoggle endpoint, and an empty string means the
  // revert flips locally with no request (an allowed-untoggle button
  // with no endpoint of its own). Left undefined, a committed click is
  // ignored: the button is sticky, which is OptimisticAction's
  // behaviour and the default of ToggleAction's.
  function bind(el, spec) {
    if (bound.has(el)) return;
    bound.add(el);
    specs.set(el, spec);
    if (spec.group) {
      let members = groups.get(spec.group);
      if (!members) {
        members = new Set();
        groups.set(spec.group, members);
      }
      members.add(el);
    }
    el.addEventListener('click', (ev) => {
      ev.preventDefault();
      const state = el.getAttribute('data-state') || 'idle';
      if (state === 'pending') return;
      // A new click cancels any rollback timer a previous error
      // scheduled. Without this, error → click again → pending → the
      // 600ms timer fires and clobbers the state back to idle while
      // the new request is still in flight.
      if (el.__fuiRollbackTimer) {
        clearTimeout(el.__fuiRollbackTimer);
        el.__fuiRollbackTimer = null;
      }
      if (state === 'committed') {
        if (spec.untoggle === undefined) return;
        setState(el, spec, 'pending');
        el.dispatchEvent(new CustomEvent('action:untoggle', { bubbles: true }));
        // No rolled-back event on a failed untoggle: the state simply
        // stays where it was (committed), because the revert was
        // refused and there is nothing to roll back to.
        const settled = spec.untoggle ? request(spec.untoggle, spec.method) : Promise.resolve(true);
        settled.then((ok) => {
          setState(el, spec, ok ? 'idle' : 'committed');
          // A refused untoggle returns this member to committed, and a
          // sibling may have committed inside the same round trip
          // (its settlement revoke skipped this one while it was
          // pending). Returning to committed displaces like any other
          // commit, so the group still converges on one member.
          if (!ok && stillGrouped(spec, el)) revokeGroupSiblings(spec.group, el);
        });
        return;
      }
      // idle or error: commit.
      setState(el, spec, 'pending');
      const revoked = spec.group ? revokeGroupSiblings(spec.group, el) : [];
      el.dispatchEvent(new CustomEvent('action:start', { bubbles: true }));
      request(spec.endpoint, spec.method).then((ok) => {
        if (ok) {
          setState(el, spec, 'committed');
          el.dispatchEvent(new CustomEvent('action:committed', { bubbles: true }));
          // Two members of one group clicked inside one round trip both
          // get here: the re-entry guard is per element and the
          // click-time revoke only sees committed siblings, so without
          // this pass both would end committed. Re-running the revoke on
          // settlement makes the last completer win and the group
          // converge on one committed member. A failure of the later
          // one still restores the sibling it displaced (below).
          if (stillGrouped(spec, el)) revokeGroupSiblings(spec.group, el);
          return;
        }
        // The server refused the new member, so the sibling it
        // displaced is still the committed one: put it back, unless
        // something else moved it or it left the page meanwhile.
        for (const other of revoked) {
          if (other.isConnected && other.getAttribute('data-state') === 'idle') {
            setState(other, specs.get(other), 'committed');
          }
        }
        setState(el, spec, 'error');
        el.dispatchEvent(new CustomEvent('action:rolled-back', { bubbles: true }));
        // After the failure settles (~600ms, the length of the shake
        // the stylesheet plays on [data-state="error"]), return to
        // idle so the button can be tried again. Stored on the element
        // so a new click can cancel it (see above); the timer firing
        // after the element left the DOM writes to a detached node,
        // which touches nothing that is on the page.
        el.__fuiRollbackTimer = setTimeout(() => {
          el.__fuiRollbackTimer = null;
          setState(el, spec, 'idle');
        }, 600);
      });
    });
  }

  // A group whose members all left the document is dropped on the
  // kernel's client-navigation event. The revoke pass prunes too, but it
  // only runs when a sibling commits; a page navigated away from
  // wholesale would otherwise be referenced by the group registry
  // forever, the last holder of every detached button on that page.
  window.addEventListener('gofastr:navigate', () => {
    for (const [name, members] of groups) {
      for (const el of members) {
        if (!el.isConnected) members.delete(el);
      }
      if (members.size === 0) groups.delete(name);
    }
  });

  // _groupCount is test-only: how many group keys the registry still
  // holds. Nothing in the module or the kernel reads it; the retention
  // tests do, so a prune that stops running fails a test instead of
  // leaking quietly.
  NS.action = { request, bind, _groupCount: () => groups.size };
  (NS.loadedModules = NS.loadedModules || {}).action = true;
})();
